agent/consul/state/catalog_events.go

```diff
@@ -0,0 +1,447 @@
+		tx := s.db.Txn(false)
+		defer tx.Abort()
+
+		connect := topic == TopicServiceHealthConnect
+		// TODO(namespace-streaming): plumb entMeta through from SubscribeRequest
+		idx, nodes, err := checkServiceNodesTxn(tx, nil, req.Key, connect, nil)
+		if err != nil {
+			return 0, err
+		}
+
+		for _, n := range nodes {
+			event := stream.Event{
+				Index: idx,
+				Topic: topic,
+				Payload: eventPayload{
+					Op:  OpCreate,
+					Obj: &n,
+				},
+			}
+
+			if n.Service != nil {
+				event.Key = n.Service.Service
+			}
+
+			// TODO: could all the events be appended as a single item?
+			buf.Append([]stream.Event{event})
+		}
+
+		return idx, err
+	}
+}
+
+type nodeServiceTuple struct {
+	Node      string
+	ServiceID string
+	EntMeta   structs.EnterpriseMeta
+}
+
+// ServiceHealthEventsFromChanges returns all the service and Connect health
+// events that should be emitted given a set of changes to the state store.
+func ServiceHealthEventsFromChanges(tx ReadTxn, changes Changes) ([]stream.Event, error) {
+	var events []stream.Event
+
+	var nodeChanges map[string]changeType
+	var serviceChanges map[nodeServiceTuple]*memdb.Change
+
+	markNode := func(node string, typ changeType) {
+		if nodeChanges == nil {
+			nodeChanges = make(map[string]changeType)
+		}
+		// If the caller has an actual node mutation ensure we store it even if the
+		// node is already marked. If the caller is just marking the node dirty
+		// without an node change, don't overwrite any existing node change we know
+		// about.
+		if nodeChanges[node] == changeIndirect {
+			nodeChanges[node] = typ
+		}
+	}
+	markService := func(node, service string, entMeta structs.EnterpriseMeta, svcChange *memdb.Change) {
+		if serviceChanges == nil {
+			serviceChanges = make(map[nodeServiceTuple]*memdb.Change)
+		}
+		k := nodeServiceTuple{
+			Node:      node,
+			ServiceID: service,
+			EntMeta:   entMeta,
+		}
+		// If the caller has an actual service mutation ensure we store it even if
+		// the service is already marked. If the caller is just marking the service
+		// dirty without an node change, don't overwrite any existing node change we
+		// know about.
+		if serviceChanges[k] == nil {
+			serviceChanges[k] = svcChange
+		}
+	}
+
+	for _, change := range changes.Changes {
+		switch change.Table {
+		case "nodes":
+			// Node changed in some way, if it's not a delete, we'll need to
+			// re-deliver CheckServiceNode results for all services on that node but
+			// we mark it anyway because if it _is_ a delete then we need to know that
+			// later to avoid trying to deliver events when node level checks mark the
+			// node as "changed".
+			n := changeObject(change).(*structs.Node)
+			markNode(n.Node, changeTypeFromChange(change))
+
+		case "services":
+			sn := changeObject(change).(*structs.ServiceNode)
+			changeCopy := change // TODO: why does the change need to be copied?
```

Comment:
I tried to change it to a value but it was failing later on because the pointer seemed to be pointing at a IndexEntry instead of a ServiceNode.

I don't see the change being modified in this case, so I'm not sure why this is breaking. It might have to do with the for loop?

Reply:
Ah yes, this is the "loop variable is re-assigned" gotcha, so taking the pointer of the original variable is not safe, because it will have the value of the last iteration.